src/System Application/App/File System/src/Account/FileAccountImpl.Codeunit.al

```diff
@@ -0,0 +1,227 @@
+        TempFileAccount.Reset();
+        TempFileAccount.DeleteAll();
+
+        foreach Connector in Connector.Ordinals do begin
+            FileSystemConnector := Connector;
+
+            FileAccounts.DeleteAll();
+            FileSystemConnector.GetAccounts(FileAccounts);
+
+            if FileAccounts.FindSet() then
+                repeat
+                    TempFileAccount := FileAccounts;
+                    TempFileAccount.Connector := Connector;
+
+                    if LoadLogos then
+                        ImportLogo(TempFileAccount, Connector);
+
+                    TempFileAccount.Insert();
+                until FileAccounts.Next() = 0;
+        end;
+
+        // Sort by account name
+        TempFileAccount.SetCurrentKey(Name);
+    end;
+
+    procedure DeleteAccounts(var FileAccountsToDelete: Record "File Account")
+    var
+        CurrentDefaultFileAccount: Record "File Account";
+        ConfirmManagement: Codeunit "Confirm Management";
+        FileScenario: Codeunit "File Scenario";
+        FileSystemConnector: Interface "File System Connector";
+    begin
+        CheckPermissions();
+
+        if not ConfirmManagement.GetResponseOrDefault(ConfirmDeleteQst, true) then
+            exit;
+
+        if not FileAccountsToDelete.FindSet() then
+            exit;
+
+        // Get the current default account to track if it was deleted
+        FileScenario.GetDefaultFileAccount(CurrentDefaultFileAccount);
+
+        // Delete all selected accounts
+        repeat
+            // Check to validate that the connector is still installed
+            // The connector could have been uninstalled by another user/session
+            if IsValidConnector(FileAccountsToDelete.Connector) then begin
+                FileSystemConnector := FileAccountsToDelete.Connector;
+                FileSystemConnector.DeleteAccount(FileAccountsToDelete."Account Id");
+            end;
+        until FileAccountsToDelete.Next() = 0;
+
+        DefaultAccountDeletion(CurrentDefaultFileAccount."Account Id", CurrentDefaultFileAccount.Connector);
+    end;
+
+    local procedure DefaultAccountDeletion(CurrentDefaultAccountId: Guid; Connector: Enum "File System Connector")
+    var
+        AllFileAccounts: Record "File Account";
+        NewDefaultFileAccount: Record "File Account";
+        FileScenario: Codeunit "File Scenario";
+    begin
+        GetAllAccounts(false, AllFileAccounts);
+
+        if AllFileAccounts.IsEmpty() then
+            exit; //All of the accounts were deleted, nothing to do
+
+        if AllFileAccounts.Get(CurrentDefaultAccountId, Connector) then
+            exit; // The default account was not deleted or it never existed
+
+        // In case there's only one account, set it as default
+        if AllFileAccounts.Count() = 1 then begin
+            MakeDefault(AllFileAccounts);
+            exit;
+        end;
+
+        Commit();  // Commit the accounts deletion in order to prompt for a new default account
```

Comment:
I dont see a delete of accounts in the code, and the table is temporary, is the commit really needed?

Reply:
Account is deleted in DeleteAccounts. This will start a transaction that needs to be commited before PromptNewDefaultAccountChoice can run a page in modal. Same in the email module.

Reply:
I would like to bring this to attention again. We should not have any unnecessary `Commit();` in the code!

Reply:
@StefanMaron I checked it also now. And the commit is needed. Isntall a Connector like the Blob Storage Connector Create two or more accounts. Delete the default one:

![image](https://github.com/user-attachments/assets/f8a579e3-f2b1-4d81-889f-c200d758bac0)

> If requesting support, please provide the following details to help troubleshooting:
> 
> Error message: 
> An error occurred and the transaction is stopped. Contact your administrator or partner for further assistance.
> 
> The following AL methods are limited during write transactions because one or more tables will be locked: Form.RunModal, Codeunit.Run, Report.RunModal, XmlPort.RunModal.
> 
> Form.RunModal is not allowed in write transactions.
> 
> Codeunit.Run is allowed in write transactions only if the return value is not used. For example, 'OK := Codeunit.Run()' is not allowed.
> 
> Report.RunModal is allowed in write transactions only if 'RequestForm = false'. For example, 'Report.RunModal(...,false)' is allowed.
> 
> XmlPort.RunModal is allowed in write transactions only if 'RequestForm = false'. For example, 'XmlPort.RunModal(...,false)' is allowed.
> 
> Use the commit method to save the changes before this call, or structure the code differently.
> 
> Contact your application developer for further assistance.
> 
> Internal session ID: 
> 447b2a23-52a9-4618-8a8d-b664b2c267b5
> 
> Application Insights session ID: 
> 993f4880-0d7d-492b-935a-aac5d9fdda8c
> 
> Client activity id: 
> ba05a984-b8fc-3740-a840-ea4410565203
> 
> Time stamp on error: 
> 2024-12-17T13:35:45.4592823Z
> 
> User telemetry id: 
> 27fc5261-36a7-469d-a24a-2e0033bc25f4
> 
> AL call stack: 
> "File Account Impl."(CodeUnit 9451).PromptNewDefaultAccountChoice line 7 - System Application by Microsoft
> "File Account Impl."(CodeUnit 9451).DefaultAccountDeletion line 21 - System Application by Microsoft
> "File Account Impl."(CodeUnit 9451).DeleteAccounts line 28 - System Application by Microsoft
> "File Accounts"(Page 9450)."Delete - OnAction"(Trigger) line 5 - System Application by Microsoft
> 
>

Reply:
Same for the Email module.
```
If requesting support, please provide the following details to help troubleshooting:

Error message: 
An error occurred and the transaction is stopped. Contact your administrator or partner for further assistance.

The following AL methods are limited during write transactions because one or more tables will be locked: Form.RunModal, Codeunit.Run, Report.RunModal, XmlPort.RunModal.

Form.RunModal is not allowed in write transactions.

Codeunit.Run is allowed in write transactions only if the return value is not used. For example, 'OK := Codeunit.Run()' is not allowed.

Report.RunModal is allowed in write transactions only if 'RequestForm = false'. For example, 'Report.RunModal(...,false)' is allowed.

XmlPort.RunModal is allowed in write transactions only if 'RequestForm = false'. For example, 'XmlPort.RunModal(...,false)' is allowed.

Use the commit method to save the changes before this call, or structure the code differently.

Contact your application developer for further assistance.

Internal session ID: 
fab8ef47-7a9c-404e-ba64-d9293fcbd1e0

Application Insights session ID: 
993f4880-0d7d-492b-935a-aac5d9fdda8c

Client activity id: 
79ecd603-1e2f-11ec-ce29-ce488f00840d

Time stamp on error: 
2024-12-17T13:46:41.9727790Z

User telemetry id: 
27fc5261-36a7-469d-a24a-2e0033bc25f4

AL call stack: 
"Email Account Impl."(CodeUnit 8889).PromptNewDefaultAccountChoice line 7 - System Application by Microsoft
"Email Account Impl."(CodeUnit 8889).HandleDefaultAccountDeletion line 25 - System Application by Microsoft
"Email Account Impl."(CodeUnit 8889).DeleteAccounts line 33 - System Application by Microsoft
"Email Account Impl."(CodeUnit 8889).DeleteAccounts line 2 - System Application by Microsoft
"Email Accounts"(Page 8887)."Delete - OnAction"(Trigger) line 5 - System Application by Microsoft
```

Reply:
Alright, and its just used on the page and can not be called from outside.
But I hope you understand why I care so much about this. If this procedure gets added to the `File Account` codeunit and gets opened up to be called from other routines, or even the internal one gets called from somewhere else inside the system app. It can happen really quickly that someone misses the Commit and unexpected behavior happens.

The email module also has a `HideDialog` implement which will prevent the `Commit`. And I think as soon as this gets added to the `File Account` codeunit it should be really well documented that there is a commit inside